test/workers/zync_worker_test.rb

```diff
@@ -44,10 +44,99 @@ class ZyncWorkerTest < ActiveSupport::TestCase
     ZyncWorker::MessageBusPublisher.expects(:enabled).returns(false)
 
     assert_difference(zync_events.method(:count), +2) do
+      worker.perform(event.event_id, event.data.with_indifferent_access)
+    end
+  end
+
+  class UnprocessableEntityRetryTest < ActiveSupport::TestCase
+    disable_transactional_fixtures!
+
+    attr_reader :worker, :event, :application
+
+    setup do
+      EventStore::Repository.stubs(raise_errors: true)
+
+      @worker = ZyncWorker.new
+      @application = FactoryBot.create(:simple_cinstance).reload # reload to get tenant_id
+      FactoryBot.create(:simple_admin, account: application.provider_account) # for the access token
+      @event = ZyncEvent.create(RailsEventStore::Event.new, application)
+
+      worker.config.stubs(endpoint: 'http://example.com') # so it makes http request
+      worker.publisher.call(event) # so it is later available to find
+
+      stub_request(:put, 'http://example.com/tenant').to_return(status: 200)
+      stub_request(:put, 'http://example.com/notification').to_return(status: 422).with(body: event.data.to_json) # causes UnprocessableEntityError to be raised
+    end
+
+    test 'sync event dependencies when it fails' do
+      worker.expects(:sync_dependencies).with(event.event_id).returns(true)
+      worker.perform(event.event_id, event.data.with_indifferent_access)
+    end
+
+    test 'raises the error when there is no dependency' do
+      worker.expects(:sync_dependencies).with(event.event_id).returns([])
       assert_raises ZyncWorker::UnprocessableEntityError do
         worker.perform(event.event_id, event.data.with_indifferent_access)
       end
     end
+
+    test '#create_dependency_events' do
+      EventStore::Repository.expects(:find_event!).returns(event)
```

Comment:
I think that we could be more specific to avoid false positives:
```ruby
EventStore::Repository.expects(:find_event!).with(event.event_id).returns(event)
```